--- sys/net/pfkeyv2.c.orig	Tue Sep 16 21:50:11 2025
+++ sys/net/pfkeyv2.c	Sat Apr 25 11:53:50 2026
@@ -102,7 +102,10 @@
 	{ SADB_X_EALG_BLF, 64, 40, BLF_MAXKEYLEN * 8},
 	{ SADB_X_EALG_CAST, 64, 40, 128},
 	{ SADB_X_EALG_AES, 128, 128, 256},
-	{ SADB_X_EALG_AESCTR, 128, 128 + 32, 256 + 32}
+	{ SADB_X_EALG_AESCTR, 128, 128 + 32, 256 + 32},
+	{ SADB_X_EALG_AESGCM16, 64, 128 + 32, 256 + 32},
+	{ SADB_X_EALG_AESGMAC, 64, 128 + 32, 256 + 32},
+	{ SADB_X_EALG_CHACHA20POLY1305, 64, 256 + 32, 256 + 32}
 };
 
 static const struct sadb_alg aalgs[] = {
