diff --git a/crypto/cipher.c b/crypto/cipher.c
index 65bcea0..bb27b14 100644
--- a/crypto/cipher.c
+++ b/crypto/cipher.c
@@ -23,6 +23,11 @@ #include <asm/scatterlist.h>
 #include "internal.h"
 #include "scatterwalk.h"
 
+static inline void xor_1(u8 *a, const u8 *b)
+{
+	a[0] ^= b[0];
+}
+
 static inline void xor_64(u8 *a, const u8 *b)
 {
 	((u32 *)a)[0] ^= ((u32 *)b)[0];
@@ -431,6 +436,11 @@ int crypto_init_cipher_ops(struct crypto
 		unsigned long addr;
 	    	
 	    	switch (crypto_tfm_alg_blocksize(tfm)) {
+		case 1:
+			/* some ciphers (e.g. null) have no restrictions. */
+			ops->cit_xor_block = xor_1;
+			break;
+			
 	    	case 8:
 	    		ops->cit_xor_block = xor_64;
 	    		break;
