From: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
Date: Wed, 25 Jan 2017 19:06:25 -0500
Subject: avoid gcc "unused variable" warnings when USE_KLIPS=false

---
 programs/pluto/nat_traversal.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/programs/pluto/nat_traversal.c b/programs/pluto/nat_traversal.c
index f78e0bf80..6dcf3cb93 100644
--- a/programs/pluto/nat_traversal.c
+++ b/programs/pluto/nat_traversal.c
@@ -729,9 +729,11 @@ void ikev1_natd_init(struct state *st, struct msg_digest *md)
 int nat_traversal_espinudp_socket(int sk, const char *fam)
 {
 	int r;
+#if defined(KLIPS)
 	struct ifreq ifr;
 	int *fdp = (int *) &ifr.ifr_data;
 	const char *ifn;
+#endif
 
 #if defined(NETKEY_SUPPORT) || defined(BSD_KAME)
 	if (kern_interface == USE_NETKEY || kern_interface == USE_BSDKAME) {
